### PR TITLE
Enable XLC 16.1 Clang-based toolchain for AIX platforms

### PR DESCRIPTION
**Pull request by vsebe**
- replace -qlanglvl=extended with -qlanglvl=extended0x
- remove unsupported options: -qinfo, -qmbcs, -qnotempinc
- add -qxlcompatmacros -fno-rtti -fno-exceptions options to C/C++ flags
- add -qlanglvl=extc99 option to *.spp rule compilation
- set xlclang++ as linker
- add xlc 16 to path and set environment variables for pipelines

Related to #6486 and https://github.com/eclipse/omr/pull/4160

[ci skip]

Signed-off-by: Violeta Sebe <vsebe@ca.ibm.com>

### CODE REVIEW

**Comment by vsebe** [comment]
@pshipton fyi

**Comment by pshipton** [comment]
Do these changes affect existing builds of Java 8 - 12, that still use xlC 13?
Some options are removed: `-qnotempinc`, `-qmbcs`, `-q mbcs`, `-qinfo=pro`. I assume these are not unsupported by xlC 13.

Also wondering if there are any side affects of switching the linker, like linking to extra dynamic libraries.

**Comment by pshipton** [comment]
@keithc-ca

**Comment by sxa555** [comment]
@pshipton @keithc-ca @vsebe Is there anything blocking this at the moment? If it's going to still be some time before it's progressed I'll disable the builds at adoptopenjdk since they are affected by this just now.

**Comment by pshipton** [comment]
@sxa555 Go ahead and disable the builds for now. This is blocked by #6486 and fixing the similar problems in OMR. Merging anything has been blocked by not being able to run eclipse ci builds until just recently. We are still determining if the ci builds are completely working, and then working through the backlog.

**Comment by sxa555** [comment]
Will do - thanks

**Comment by pshipton** [comment]
jenkins test sanity aix jdk8,jdk13 depends eclipse/omr#4160

**Comment by pshipton** [comment]
We're missing a change to the defaults.yml to set xlc 16. Not sure if xlc 16 is installed on the OpenJ9 machines yet.
```
13: 'PATH+XLC=/opt/IBM/xlC/16.1.0/bin:/opt/IBM/xlc/16.1.0/bin CC=xlclang CXX=xlclang++
next: 'PATH+XLC=/opt/IBM/xlC/16.1.0/bin:/opt/IBM/xlc/16.1.0/bin CC=xlclang CXX=xlclang++

**Comment by pshipton** [comment]
Latest problem is
```
21:12:20  In file included from /home/jenkins/workspace/Build_JDK13_ppc64_aix_Personal/build/aix-ppc64-server-release/vm/compiler/../omr/compiler/ras/CallStack.cpp:44:
21:12:20  /usr/include/demangle.h:137:2: error: "Please include ibmdemangle.h for xlclang++ to demangle symbol names."
21:12:20  #error "Please include ibmdemangle.h for xlclang++ to demangle symbol names."
21:12:20   ^
21:12:20  8 warnings and 1 error generated.
```
@gita-omr

**Comment by gita-omr** [comment]
Sent a note to the xlC team regarding ibmdemangle.h

**Comment by DanHeidinga** [comment]
@gita-omr Is the demangle functionality critical for the JIT to run?  If its only used for `ras` purposes (ie: native call stacks in javacores), can it be temporarily disabled so we can make progress on bringing JDK13 up on AIX?

**Comment by gita-omr** [comment]
@DanHeidinga I suspect we can, but I am away most of next week. Could you please work with @zl-wang on this?

**Comment by DanHeidinga** [comment]
It would be great to have @zl-wang involved in this.  It makes more sense to have someone working on POWER helping move this forward than pulling me in.  Can you work with @zl-wang and other PPC devs to move this forward?

**Comment by zl-wang** [comment]
The fix is anticipated to come sometime this week.

File: buildspecs/aix_ppc-64_cmprssptrs.spec [hunk 1]
```diff
@@ -91,8 +91,8 @@ SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-excepti
 		<property name="uma_make_cmd_cc" value="xlc_r"/>
 		<property name="uma_make_cmd_cpp" value="$(CC) -P"/>
 		<property name="uma_make_cmd_cxx" value="xlC_r"/>
-		<property name="uma_make_cmd_cxx_dll_ld" value="$(CC)"/>
-		<property name="uma_make_cmd_cxx_exe_ld" value="$(CC)"/>
+		<property name="uma_make_cmd_cxx_dll_ld" value="$(CXX)"/>
+		<property name="uma_make_cmd_cxx_exe_ld" value="$(CXX)"/>
```
Comment (keithc-ca) [inline buildspecs/aix_ppc-64_cmprssptrs.spec]: Assuming this change is reasonable, we'll also need to change `aix_ppc-64.spec` similarly.

File: runtime/makelib/targets.mk.aix.inc.ftl [hunk 2]
```diff
@@ -92,15 +92,31 @@ ifdef I5_VERSION
 else
   CFLAGS += -s $(UMA_CC_MODE)
 endif
-CFLAGS += -q mbcs -qlanglvl=extended -qarch=ppc -qinfo=pro -qalias=noansi -qxflag=LTOL:LTOL0 -qsuppress=1506-1108
+
+ifeq ($(notdir $(CC)),xlclang)
```
Comment (keithc-ca) [inline runtime/makelib/targets.mk.aix.inc.ftl]: This pattern assumes that `CC` is not something like `xlclang-16`.

File: buildenv/jenkins/variables/defaults.yml [hunk 3]
```diff
@@ -244,7 +244,7 @@ ppc64_aix:
       11: 'PATH+XLC=/opt/IBM/xlC/13.1.3/bin:/opt/IBM/xlc/13.1.3/bin'
       12: 'PATH+XLC=/opt/IBM/xlC/13.1.3/bin:/opt/IBM/xlc/13.1.3/bin'
       13: 'PATH+XLC=/opt/IBM/xlC/13.1.3/bin:/opt/IBM/xlc/13.1.3/bin'
```
Comment (pshipton) [inline buildenv/jenkins/variables/defaults.yml]: 13 needs the same change as next

File: runtime/makelib/targets.mk.aix.inc.ftl (hunk 2 above)
Comment (pshipton) [reply runtime/makelib/targets.mk.aix.inc.ftl]: The statement was updated, @keithc-ca any further concerns?

**Comment by keithc-ca** [comment]
I don't see an answer to this question from @pshipton:
> Also wondering if there are any side affects of switching the linker, like linking to extra dynamic libraries.

File: buildenv/jenkins/variables/defaults.yml (hunk 3 above)
Comment (vsebe) [reply buildenv/jenkins/variables/defaults.yml]: fixed

**Comment by zl-wang** [comment]
Pending performance evaluation of xlclang build, comparing to xlC build. As far as I know, xlclang has never been intended for performance.

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk [hunk 4]
```diff
@@ -114,6 +112,20 @@ ifeq ($(BUILD_CONFIG),prod)
     CX_FLAGS+=$(CX_FLAGS_PROD)
 endif
 
+ifeq (,$(findstring xlclang,$(notdir $(CC))))
+  # xlc options
+  CX_FLAGS+=-qnotempinc -qmbcs
+else
+  # xlclang
+  CX_FLAGS+=-qxlcompatmacros
+  SPP_FLAGS+=-qlanglvl=extc99
```
Comment (pshipton) [inline runtime/compiler/build/toolcfg/aix-xlc/common.mk]: @vsebe do you have a record for why this option was needed?

File: runtime/makelib/targets.mk.aix.inc.ftl (hunk 2 above)
Comment (keithc-ca) [reply runtime/makelib/targets.mk.aix.inc.ftl]: Given it appears that xlclang alters its behavior based on the name used to invoke the compiler, this seems reasonable (xlc and xlclang use the same binary).

**Comment by pshipton** [comment]
> I don't see an answer to this question 

Elsewhere, I've asked for the output from `ldd libj9vm29.so` and `ldd libj9jit29.so` for both the original and new builds.

**Comment by pshipton** [comment]
> Pending performance evaluation of xlclang build, comparing to xlC build. As far as I know, xlclang has never been intended for performance.

@zl-wang note these changes only affect Java 13 and later. Not sure we have a choice, as OpenJDK seems to have removed support for using xlC. https://bugs.openjdk.java.net/browse/JDK-8224214

**Comment by zl-wang** [comment]
My point is  OpenJDK side has no imperative reason to switch to xlclang.  It looked to me the switch was based on xlclang being newer in the pack of compiler commands. If we can prove it is say more than 10% worse than xlC build (jit compiler and GC mainly), we should ask them to reverse.

**Comment by pshipton** [comment]
> If we can prove it is say more than 10% worse than xlC build (jit compiler and GC mainly), we should ask them to reverse.

Sure. I also note that xlclang is recommended as follows, which may have been a factor in switching.
https://www.ibm.com/support/knowledgecenter/en/SSGH3R_16.1.0/com.ibm.xlcpp161.aix.doc/getstart/new_xlclang.html
```
Compared with the legacy XL-based front end, the new Clang-based front end provides the following capabilities:

    Enhanced language standard support:

    If your programs conform to a newer language standard, like the C++11 or C++14 language standard, which is not fully supported by the legacy XL-based front end, you might want to compile the programs with the new Clang-based front end.
```

**Comment by zl-wang** [comment]
Yes, better support for language standard is all the value proposal for xlclang.  Performance is not. Let's wait and see where we end up.

**Comment by pshipton** [comment]
jenkins test sanity,extended aix jdk13 depends eclipse/omr#4160

**Comment by pshipton** [comment]
Forgot the demangle problem is still pending, but at least we can get that far in a PR build now.
```
In file included from /home/u0020236/workspace/Build_JDK13_ppc64_aix_Personal/build/aix-ppc64-server-release/vm/compiler/../omr/compiler/ras/CallStack.cpp:44:
13:27:17  /usr/include/demangle.h:137:2: error: "Please include ibmdemangle.h for xlclang++ to demangle symbol names."
13:27:17  #error "Please include ibmdemangle.h for xlclang++ to demangle symbol names."
```

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk (hunk 4 above)
Comment (vsebe) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: This is related to string concatenation - see also https://github.com/eclipse/openj9/pull/6480.
extc99 is one of the supported defaults for xlclang (as per https://www.ibm.com/support/knowledgecenter/SSGH3R_16.1.0/com.ibm.xlcpp161.aix.doc/compiler_ref/opt_langlvl_aix.html)

**Comment by pshipton** [comment]
jenkins test sanity aix jdk8,jdk11 depends eclipse/omr#4160

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk (hunk 4 above)
Comment (keithc-ca) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: The illegal code was fixed: do we still need `-qlanglvl=extc99`?

Comment (vsebe) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: yes, we do

Comment (keithc-ca) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: Please explain why we need that option. Where & how does it make a difference?

Comment (vsebe) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: It is require to preprocess ebb.spp, e.g. `xclang -qlanglvl=extc99 ... -E openj9/runtime/compiler/p/runtime/ebb.spp`

**Comment by midronij** [comment]
@vsebe @pshipton Here I my changes for the demangling issue: __. Apologies for the delay.

When I ran a personal build, everything compiled perfectly fine, but at the linking stage, I got this error:
```
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_guard_acquire
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_guard_release
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_end_catch
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__xlc_exception_handle
20:41:41  ld: 0711-317 ERROR: Undefined symbol: ._Unwind_Resume
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_begin_catch
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_guard_abort
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_rethrow
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::terminate()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::unexpected()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_allocate_exception
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::bad_alloc::bad_alloc()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: vtable for __cxxabiv1::__si_class_type_info
20:41:41  ld: 0711-317 ERROR: Undefined symbol: vtable for __cxxabiv1::__class_type_info
20:41:41  ld: 0711-317 ERROR: Undefined symbol: std::bad_alloc::~bad_alloc()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_throw
20:41:41  ld: 0711-317 ERROR: Undefined symbol: __cxa_pure_virtual
20:41:41  ld: 0711-317 ERROR: Undefined symbol: std::nothrow
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__xlc_catch_matchv2
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::__1::__vector_base_common<true>::__throw_length_error() const
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::exception::~exception()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: vtable for __cxxabiv1::__vmi_class_type_info
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .__cxa_demangle
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::bad_alloc::bad_alloc()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::bad_alloc::~bad_alloc()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::logic_error::logic_error(char const*)
20:41:41  ld: 0711-317 ERROR: Undefined symbol: vtable for std::length_error
20:41:41  ld: 0711-317 ERROR: Undefined symbol: std::length_error::~length_error()
20:41:41  ld: 0711-317 ERROR: Undefined symbol: .std::__1::__vector_base_common<true>::__throw_out_of_range() const
20:41:41  ld: 0711-317 ERROR: Undefined symbol: vtable for std::exception
20:41:41  ld: 0711-344 See the loadmap file lmap for more information.
```
Here's the link to the build if you want to see the full console log: https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK13_ppc64_aix_Personal/13/console

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk (hunk 4 above)
Comment (keithc-ca) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: It was required before #6480 was merged, but is it *still* required?

**Comment by DanHeidinga** [comment]
@midronij That looks like the code is expecting to be linked with libstdc++ (or similar named c++ library).  In OpenJ9, the JIT statically links the c++ std library so this may be a compilation flag issue or a link time issue. 

Can you provide the complete compile and link lines for the file using demangle?

File: runtime/compiler/build/toolcfg/aix-xlc/common.mk (hunk 4 above)
Comment (vsebe) [reply runtime/compiler/build/toolcfg/aix-xlc/common.mk]: Yes, it is still required.

**Comment by midronij** [comment]
@DanHeidinga This was the link command used ``` makeC++SharedLib_r -p0 -bloadmap:lmap -brtl -bnoentry -bnolibpath -X64   -L/home/jenkins/workspace/Build_JDK13_ppc64_aix_Personal/build/aix-ppc64-server-release/vm/compiler/.. -L/home/jenkins/workspace/Build_JDK13_ppc64_aix_Personal/build/aix-ppc64-server-release/vm/compiler/../lib -o ``` I spoke to someone on the xlC team about this, and apparently makeC++SharedLib_r not including libstdc++ is a bug. I've tried changing the linker to xlclang++, and if that fails, I can hard code it in as a temporary solution, to verify that this is indeed the cause of the problem.

**Comment by zl-wang** [comment]
@midronij @DanHeidinga    I was told that makeSharedLib_r is not going to add -lc++ automatically, since it is used for linking legacy object files too which don't depend on -lc++  (which is different from stdc++).

**Comment by midronij** [comment]
@vsebe @pshipton I took care of the -lc++ issue (I just added it in manually), but the build fails later on. This was the error message: 
```
13:01:53  Running ddrgen to generate j9ddr.dat and superset.dat

13:03:01  Error: dwarf/DwarfScanner.cpp:613 getTypeTag - Getting tag from type die: DW_DLE_DIE_NULL (3)
13:03:01  
13:03:01  Error: dwarf/DwarfScanner.cpp:1507 traverse_cu_in_debug_section - Failed to add type.
13:03:01  

13:03:02  Error: dwarf/DwarfScanner.cpp:1538 startScan - Failure scanning ../libj9gc29.so.dbg
```
The full console log output can be seen here: https://hyc-runtimes-jenkins.swg-devops.com/job/Build_JDK13_ppc64_aix_Personal/24/console

**Comment by DanHeidinga** [comment]
@keithc-ca Any suggestions for dealing with the DWARF effors in https://github.com/eclipse/openj9/pull/6584#issuecomment-530050416 ?

**Comment by midronij** [comment]
Here are the PRs for the changes I've made so far:
https://github.com/eclipse/omr/pull/4285
https://github.com/eclipse/openj9/pull/7034

**Comment by pshipton** [comment]
I'm going to go ahead and merge this to help move AIX xlC 16 compiles along. It's compatible with older java versions, and is part of what is needed for xlC 16. There are and will be other PRs created to resolve other problems.

**Comment by pshipton** [comment]
@vsebe please create a PR for this against the v0.16.0-release branch

**Comment by vsebe** [comment]
@pshipton please see https://github.com/eclipse/openj9/pull/7039

**Pull request merged**